register-machines/src/main.rs

```diff
@@ -11,55 +11,203 @@ enum Instruction {
     Halt
 }
 
-use Instruction::*;
-fn eval_program(program: &[Instruction], state: &State) -> State {
-    unimplemented!();
+type Godel = u128;
+
+trait Encodable {
+    fn encode(&self) -> Godel;
 }
-// <<x,y>> = (2^x)*(2y+1)
-fn encode_pair1(x: u128, y: u128) -> u128 {
-    unimplemented!();
+
+impl<T: Encodable> Encodable for &T {
+    fn encode(&self) -> Godel {
+        (**self).encode()
+    }
 }
-// <x,y> = (2^x)*(2y+1)-1
-fn encode_pair2(x: u128, y: u128) -> u128 {
-    unimplemented!();
+
+impl Encodable for u128 {
+    fn encode(&self) -> Godel {
+        *self
+    }
 }
-fn encode_list_to_godel(l: &[u128]) -> u128 {
-    unimplemented!();
+
+impl Encodable for i32 {
+    fn encode(&self) -> Godel {
+        *self as u128
+    }
 }
-fn encode_program_to_list(program: &[Instruction]) -> Vec<u128> {
-    unimplemented!();
+
+impl Encodable for usize {
+    fn encode(&self) -> Godel {
+        *self as u128
+    }
 }
-// a = (2^x)*(2y+1)
-fn decode_pair1(a: u128) -> (u128, u128) {
-    unimplemented!();
+
+impl<T: Encodable, V: Encodable> Encodable for (T, V) {
+    fn encode(&self) -> Godel {
+        let (x, y) = self;
+        (1 << x.encode()) * (2 * y.encode() + 1) - 1
+    }
 }
-// a = (2^x)*(2y+1)-1
-fn decode_pair2(a: u128) -> (u128, u128) {
-    unimplemented!();
+
+impl<T: Encodable> Encodable for Option<T> {
+    fn encode(&self) -> Godel {
+        match self {
+            Some(x) => 1 + x.encode(),
+            None => 0
+        }
+    }
 }
-fn decode_godel_to_list(g: u128) -> Vec<u128> {
-    unimplemented!();
+
+impl Encodable for Instruction {
+    fn encode(&self) -> Godel {
+        match *self {
+            Add(i, j) => Some((2 * i, j)).encode(),
+            Sub(i, j, k) => Some((2 * i + 1, (j, k))).encode(),
+            Halt => (None as Option<u128>).encode()
+        }
+    }
+}
+
+impl<T: Encodable> Encodable for [T] {
+    fn encode(&self) -> Godel {
+        self.iter()
+            .rev()
+            .fold(
+                (None as Option<u128>).encode(),
+                |x, y| Some((y, x)).encode()
+            )
+    }
+}
```

Comment:
Brilliant use of an almost completely generic function. When I could get it to compile to assembly (a lot harder than you might think when optimisation gets rid of unused code) it showed:
```
example::encode:
        test    rsi, rsi
        je      .LBB12_1
        shl     rsi, 4
        add     rdi, rsi
        add     rsi, -16
        mov     r9, rsi
        shr     r9, 4
        add     r9, 1
        mov     r8d, r9d
        and     r8d, 3
        cmp     rsi, 48
        jae     .LBB12_4
        xor     eax, eax
        xor     edx, edx
        jmp     .LBB12_6
.LBB12_1:
        xor     eax, eax
        xor     edx, edx
        ret
```
Which is nice and optimised, very cool :).